File: taskhub/lib/models/company.dart
import 'package:taskhub/models/citizen.dart';
import 'package:taskhub/models/professional.dart';

class Company {
  final int? id;
  final int userId;
  final String companyName;
  final String? legalName;
  final String cnpj;
  final String? registrationNumber;
  final String serviceType;
  final String? specialties;
  final String? description;
  final int? employeeCount;
  final String? portfolioUrl;
  final List<String>? portfolioItems;
  final List<Certification>? certifications;
  final String? websiteUrl;
  final int? workStartedYear;
  final Address? address;
  final Map<String, dynamic>? businessHours;
  final double? rating;
  final int? totalReviews;

  Company({
    this.id,
    required this.userId,
    required this.companyName,
    this.legalName,
    required this.cnpj,
    this.registrationNumber,
    required this.serviceType,
    this.specialties,
    this.description,
    this.employeeCount,
    this.portfolioUrl,
    this.portfolioItems,
    this.certifications,
    this.websiteUrl,
    this.workStartedYear,
    this.address,
    this.businessHours,
    this.rating,
    this.totalReviews,
  });

  int get yearsInBusiness {
    if (workStartedYear == null) return 0;
    return DateTime.now().year - workStartedYear!;
  }

  String get ratingDisplay {
    if (rating == null || rating == 0) return 'Sem avaliações';
    return '${rating!.toStringAsFixed(1)} ⭐ ($totalReviews avaliações)';
  }

  factory Company.fromJson(Map<String, dynamic> json) {
    List<Certification>? certificationsList;
    if (json['certifications'] != null) {
      certificationsList = (json['certifications'] as List)
          .map((c) => Certification.fromJson(c))
          .toList();
    }

    return Company(
      id: json['id'],
      userId: json['user_id'] ?? 0,
      companyName: json['company_name'] ?? '',
      legalName: json['legal_name'],
      cnpj: json['cnpj'] ?? '',
      registrationNumber: json['registration_number'],
      serviceType: json['service_type'] ?? '',
      specialties: json['specialties'],
      description: json['description'],
      employeeCount: json['employee_count'],
      portfolioUrl: json['portfolio_url'],
      portfolioItems: json['portfolio_items'] != null
          ? List<String>.from(json['portfolio_items'])
          : null,
      certifications: certificationsList,
      websiteUrl: json['website_url'],
      workStartedYear: json['work_started_year'],
      address: json['address'] != null
          ? Address.fromJson(json['address'])
          : null,
      businessHours: json['business_hours'],
      rating: json['rating'] != null
          ? double.tryParse(json['rating'].toString())
          : null,
      totalReviews: json['total_reviews'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'user_id': userId,
      'company_name': companyName,
      'legal_name': legalName,
      'cnpj': cnpj,
      'registration_number': registrationNumber,
      'service_type': serviceType,
      'specialties': specialties,
      'description': description,
      'employee_count': employeeCount,
      'portfolio_url': portfolioUrl,
      'portfolio_items': portfolioItems,
      'certifications': certifications?.map((c) => c.toJson()).toList(),
      'website_url': websiteUrl,
      'work_started_year': workStartedYear,
      'address': address?.toJson(),
      'business_hours': businessHours,
    };
  }
}

File: backend/src/models/Company.js
class Company {
    constructor({
        id,
        user_id,
        company_name,
        legal_name,
        cnpj,
        registration_number,
        service_type,
        specialties,
        description,
        employee_count,
        portfolio_url,
        portfolio_items,
        certifications,
        website_url,
        work_started_year,
        address_street,
        address_number,
        address_complement,
        address_neighborhood,
        address_city,
        address_state,
        address_zip_code,
        address_country,
        business_hours,
        rating,
        total_reviews,
        created_at,
        updated_at,
    }) {
        this.id = id;
        this.user_id = user_id;
        this.company_name = company_name;
        this.legal_name = legal_name;
        this.cnpj = cnpj;
        this.registration_number = registration_number;
        this.service_type = service_type;
        this.specialties = specialties;
        this.description = description;
        this.employee_count = employee_count;
        this.portfolio_url = portfolio_url;
        this.portfolio_items = typeof portfolio_items === 'string' ? JSON.parse(portfolio_items || '[]') : portfolio_items || [];
        this.certifications = typeof certifications === 'string' ? JSON.parse(certifications || '[]') : certifications || [];
        this.website_url = website_url;
        this.work_started_year = work_started_year;
        this.address_street = address_street;
        this.address_number = address_number;
        this.address_complement = address_complement;
        this.address_neighborhood = address_neighborhood;
        this.address_city = address_city;
        this.address_state = address_state;
        this.address_zip_code = address_zip_code;
        this.address_country = address_country;
        this.business_hours = typeof business_hours === 'string' ? JSON.parse(business_hours || '{}') : business_hours || {};
        this.rating = rating;
        this.total_reviews = total_reviews;
        this.created_at = created_at;
        this.updated_at = updated_at;
    }

    /**
     * Get full address
     */
    getFullAddress() {
        const parts = [
            this.address_street,
            this.address_number,
            this.address_complement,
            this.address_neighborhood,
            this.address_city,
            `${this.address_state} ${this.address_zip_code}`,
            this.address_country,
        ].filter(Boolean);
        return parts.join(', ');
    }

    /**
     * Get years in business
     */
    getYearsInBusiness() {
        if (!this.work_started_year) return 0;
        return new Date().getFullYear() - this.work_started_year;
    }

    /**
     * Get rating display
     */
    getRatingDisplay() {
        if (!this.rating) return 'Sem avaliações';
        return `${this.rating.toFixed(1)} ⭐ (${this.total_reviews} avaliações)`;
    }

    /**
     * Convert to JSON for response
     */
    toJSON() {
        return {
            id: this.id,
            user_id: this.user_id,
            company_name: this.company_name,
            legal_name: this.legal_name,
            cnpj: this.cnpj,
            registration_number: this.registration_number,
            service_type: this.service_type,
            specialties: this.specialties,
            description: this.description,
            employee_count: this.employee_count,
            portfolio_url: this.portfolio_url,
            portfolio_items: this.portfolio_items,
            certifications: this.certifications,
            website_url: this.website_url,
            work_started_year: this.work_started_year,
            years_in_business: this.getYearsInBusiness(),
            address: {
                street: this.address_street,
                number: this.address_number,
                complement: this.address_complement,
                neighborhood: this.address_neighborhood,
                city: this.address_city,
                state: this.address_state,
                zip_code: this.address_zip_code,
                country: this.address_country,
            },
            business_hours: this.business_hours,
            rating: this.rating,
            total_reviews: this.total_reviews,
        };
    }
}

module.exports = Company;
